src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -388,7 +388,8 @@ private void closeShardInjector(String reason, ShardId sId, Injector shardInject
             // and close the shard so no operations are allowed to it
             if (indexShard != null) {
                 try {
-                    indexShard.close(reason);
+                    final boolean flushEngine = deleted.get() == false && closed.get(); // only flush we are we closed (closed index or shutdown) and if we are not deleted
```

Comment:
+1 on the closed indices use case. Good catch. I'm not sure it has effect now. Did some sniffing and we close the shards in too many places upon index close. First place in `IndicesClusterStateService.applyDeletedShards`, then we have another iteration in the beginning of `IndicesClusterStateService.applyCleanedIndices` : 

```
for (IndexService indexService : indicesService) {
            String index = indexService.index().getName();
            IndexMetaData indexMetaData = event.state().metaData().index(index);
            if (indexMetaData != null && indexMetaData.state() == IndexMetaData.State.CLOSE) {
                for (Integer shardId : indexService.shardIds()) {
                    logger.debug("[{}][{}] removing shard (index is closed)", index, shardId);
                    try {
                        indexService.removeShard(shardId, "removing shard (index is closed)");
                    } catch (Throwable e) {
                        logger.warn("[{}] failed to remove shard (index is closed)", e, index);
                    }
                }
            }
        }
```

and only then we do `indicesService.removeIndex(index, reason);` which closes the index (but it has no shards any more..)